cylc/uiserver/handlers.py

```diff
@@ -13,109 +13,153 @@
+        #   removed rrom the URL for subsequent requests, the cookie is
+        #   used in its place.
+        # * If the token was used token_authenticated is True.
+        # * If the cookie was used it is False (despite the cookie auth
+        #   being derived from token auth).
+        # * Due to a bug in jupyter_server the user is returned as bytes
+        #   when cookie auth is used so at present we can use this to
+        #   tell.
+        #   https://github.com/jupyter-server/jupyter_server/pull/562
+        # TODO: this hack is obviously not suitable for production!
+        return True
+    elif handler.token_authenticated:
+        # standalone UIS, the bearer of the token is the owner
+        # (no multi-user functionality so no futher auth required)
+        return True
+    return False
+
+
+def _authorise(
+    handler: 'CylcAppHandler',
+    username: str,
+    action: str = 'READ'
+) -> bool:
+    """Authorises a user to perform an action.
+
+    Currently this returns False unless the authenticated user is the same
+    as the user this server is running under.
+    """
+    if username != ME:
+        # auth provided by the hub, check the user name
+        handler.log.warning(f'Authorisation failed for {username}')
+        return False
+    return True
 
-class BaseHandler(HubOAuthenticated, web.RequestHandler):
 
-    def set_default_headers(self) -> None:
-        self.set_header("X-JupyterHub-Version", jupyterhub_version)
-        self.set_header("Access-Control-Allow-Origin", "*")
-        self.set_header("Access-Control-Allow-Headers", "x-requested-with")
-        self.set_header('Access-Control-Allow-Methods', 'POST, GET, OPTIONS')
-        # prevent server fingerprinting
-        self.clear_header('Server')
+class CylcAppHandler(JupyterHandler):
+    """Base handler for Cylc endpoints.
 
+    This handler adds the Cylc authorisation layer which is triggered by
+    calling CylcAppHandler.get_current_user which is called by
+    web.authenticated.
 
-class StaticHandler(BaseHandler, web.StaticFileHandler):
-    """A static handler that extends BaseHandler (for headers)."""
+    When running as a Hub application the make_singleuser_app method patches
+    this handler to insert the HubOAuthenticated bases class high up
+    in the inheritance order.
 
+    https://github.com/jupyterhub/jupyterhub/blob/3800ceaf9edf33a0171922b93ea3d94f87aa8d91/jupyterhub/singleuser/mixins.py#L826
+    """
 
-class MainHandler(BaseHandler):
+    auth_level = None
 
-    # hub_users = ["kinow"]
-    # hub_groups = []
-    # allow_admin = True
+    @property
+    def hub_users(self):
+        # allow all users (handled by Cylc authorisation decorator)
+        return None
 
-    def initialize(self, path):
-        self._static = path
+    @property
+    def hub_groups(self):
+        # allow all groups (handled by Cylc authorisation decorator)
+        return None
```

Comment:
This effectively deactivates the default Jupyter Server authorisation allowing us to open up these endpoints. All secure endpoints MUST have `@authorised` decorators.